File: tests/r_parity/44_mlogit.py
"""StatsPAI multinomial logit parity (Python side) -- Module 44."""
from __future__ import annotations
import numpy as np, pandas as pd, statspai as sp
from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "44_mlogit"

def make_data(n=1000, seed=PARITY_SEED):
    rng = np.random.default_rng(seed)
    x1 = rng.normal(0, 1, n); x2 = rng.normal(0, 1, n)
    u1 = 0.3 + 0.5 * x1 - 0.2 * x2
    u2 = -0.1 - 0.3 * x1 + 0.4 * x2
    expu = np.column_stack([np.zeros(n), u1, u2])
    expu = np.exp(expu - expu.max(axis=1, keepdims=True))
    p = expu / expu.sum(axis=1, keepdims=True)
    y = np.array([rng.choice(3, p=p[i]) for i in range(n)])
    return pd.DataFrame({"y": y, "x1": x1, "x2": x2})

def main():
    df = make_data()
    dump_csv(df, MODULE)
    res = sp.mlogit(formula="y ~ x1 + x2", data=df, base=0)
    rows = []
    # sp parameter naming: [<class>]<varname>, e.g. "[1]x1", "[1]_cons"
    for cls in [1, 2]:
        for nm, label in [("_cons", "intercept"), ("x1", "x1"), ("x2", "x2")]:
            key = f"[{cls}]{nm}"
            if key in res.params.index:
                rows.append(ParityRecord(MODULE, "py",
                    f"class{cls}_{label}",
                    estimate=float(res.params[key]),
                    se=float(res.std_errors[key]),
                    n=int(len(df))))
    write_results(MODULE, "py", rows, extra={"base_class": 0, "engine": "statsmodels"})

if __name__ == "__main__":
    main()

File: tests/r_parity/44_mlogit.R
# StatsPAI multinomial logit parity (R side) -- Module 44.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) dirname(normalizePath(sub("^--file=", "", .file_arg[1]))) else getwd()
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({ library(nnet) })

MODULE <- "44_mlogit"
df <- read_csv_strict(MODULE)
df$y <- factor(df$y)

fit <- nnet::multinom(y ~ x1 + x2, data = df, trace = FALSE, maxit = 200)

co <- summary(fit)$coefficients     # rows: classes 1, 2; cols: intercept, x1, x2
se <- summary(fit)$standard.errors

rows <- list()
for (cls in c("1", "2")) {
  for (term in c("(Intercept)", "x1", "x2")) {
    label <- if (term == "(Intercept)") "intercept" else term
    rows[[length(rows) + 1L]] <- parity_row(
      MODULE, paste0("class", cls, "_", label),
      estimate = unname(co[cls, term]),
      se = unname(se[cls, term]),
      n = nrow(df))
  }
}

write_results(MODULE, rows, extra = list(base_class = 0, package = "nnet::multinom"))
